src/Pos/Util.hs

```diff
@@ -0,0 +1,377 @@
+
+-- | Lens for the last element of 'NonEmpty'.
+_neLast :: Lens' (NonEmpty a) a
+_neLast f (x :| []) = (:| []) <$> f x
+_neLast f (x :| xs) = (\y -> x :| unsafeInit xs ++ [y]) <$> f (unsafeLast xs)
+
+----------------------------------------------------------------------------
+-- LRU cache
+----------------------------------------------------------------------------
+
+-- | Remove all items from LRU, retaining maxSize property.
+clearLRU :: Ord k => LRU.LRU k v -> LRU.LRU k v
+clearLRU = LRU.newLRU . LRU.maxSize
+
+-- [SRK-51]: Probably this instance should be in @safecopy@ library
+instance (Ord k, SafeCopy k, SafeCopy v) =>
+         SafeCopy (LRU.LRU k v) where
+    getCopy = contain $ LRU.fromList <$> safeGet <*> safeGet
+    putCopy lru =
+        contain $
+        do safePut $ LRU.maxSize lru
+           safePut $ LRU.toList lru
+    errorTypeName _ = "LRU"
+
+----------------------------------------------------------------------------
+-- Deserialized wrapper
+----------------------------------------------------------------------------
+
+eitherToVerRes :: Either Text a -> VerificationRes
+eitherToVerRes (Left errors) = if T.null errors then VerFailure []
+                               else VerFailure $ T.split (==';') errors
+eitherToVerRes (Right _ )    = VerSuccess
+
+
+instance IsString s => MonadFail (Either s) where
+    fail = Left . fromString
+
+instance MonadFail (ParsecT s u m) where
+    fail = Monad.fail
+
+deriving instance MonadFail m => MonadFail (LoggerNameBox m)
+
+instance MonadFail m => MonadFail (ResourceT m) where
+    fail = lift . fail
+
+----------------------------------------------------------------------------
+-- Concurrency utilites (MVar/TVar/RWLock..)
+----------------------------------------------------------------------------
+
+clearMVar :: MonadIO m => MVar a -> m ()
+clearMVar = liftIO . void . tryTakeMVar
+
+forcePutMVar :: MonadIO m => MVar a -> a -> m ()
+forcePutMVar mvar val = do
+    res <- liftIO $ tryPutMVar mvar val
+    unless res $ do
+        _ <- liftIO $ tryTakeMVar mvar
+        forcePutMVar mvar val
+
+-- | Block until value in MVar satisfies given predicate. When value
+-- satisfies, it is returned.
+readMVarConditional :: (MonadIO m) => (x -> Bool) -> MVar x -> m x
+readMVarConditional predicate mvar = do
+    rData <- liftIO . readMVar $ mvar -- first we try to read for optimization only
+    if predicate rData then pure rData
+    else do
+        tData <- liftIO . takeMVar $ mvar -- now take data
+        if predicate tData then do -- check again
+            _ <- liftIO $ tryPutMVar mvar tData -- try to put taken value
+            pure tData
+        else
+            readMVarConditional predicate mvar
+
+-- | Read until value is equal to stored value comparing by some function.
+readUntilEqualMVar
+    :: (Eq a, MonadIO m)
+    => (x -> a) -> MVar x -> a -> m x
+readUntilEqualMVar f mvar expVal = readMVarConditional ((expVal ==) . f) mvar
```

Comment:
Both `readMVarConditional` and `readUntilEqualMVar` are both unused, which is good because the former seems like a bad idea.